commit 4ba953577be8765a90228cc33d7fcc1dd232a23b (refs/changes/73/2373/7)
Author: Kelly McLaughlin <kelly.mclaughlin@joyent.com>
Date:   2017-09-08T15:06:12-06:00 (2 years, 1 month ago)
    
    MANTA-3338 muskie leaked a connection
    Reviewed by: David Pacheco <dap@joyent.com>
    Approved by: Jared Morrow <jm@joyent.com>

diff --git a/lib/server.js b/lib/server.js
index a1e58a6..b04eb4b 100644
--- a/lib/server.js
+++ b/lib/server.js
@@ -426,7 +426,20 @@ function createServer(options, clearProxy) {
         _audit(req, res, route, err);
     });
 
-    server.on('after', _audit);
+    server.on('after', function (req, res, route, err) {
+        _audit(req, res, route, err);
+
+        if ((req.method === 'PUT' || req.method === 'POST') &&
+            res.statusCode >= 500) {
+            /*
+             * An error occurred on a PUT or POST request, but there may still
+             * be incoming data on the request stream. Call resume() in order to
+             * dump any remaining request data so the stream emits an 'end' and
+             * the socket resources are not leaked.
+             */
+            req.resume();
+        }
+    });
 
     return (server);
 }
